src/Sylius/Bundle/CoreBundle/Resources/config/services.xml

```diff
@@ -265,7 +265,12 @@
             id="Sylius\Component\Core\Provider\CatalogPromotionVariantsProviderInterface"
             class="Sylius\Bundle\CoreBundle\Provider\CatalogPromotionVariantsProvider"
         >
+            <argument type="tagged_iterator" tag="sylius.catalog_promotion.variants_provider" />
+        </service>
+
+        <service id="Sylius\Bundle\CoreBundle\Provider\ForVariantsRuleVariantsProvider">
             <argument type="service" id="sylius.repository.product_variant" />
+            <tag name="sylius.catalog_promotion.variants_provider" />
```

Comment:
Btw, we should probably mention this tag somewhere in the documentation 🖖

Reply:
absolutely we should, but IMHO it should be in separate PR (with an explanation of how it works with more rules), I will add subtask to our board :)